### PR TITLE
linear_cross_entropy: bound chunked peak memory with the budget chunking method

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #187053
* __->__ #187271

The chunked scalar op precomputes grad_input/grad_weight in forward, so at
its peak it holds both inputs and both gradients (2*(N+V)*in_features*
itemsize) plus the per-chunk working buffers. The aspect_ratio heuristic
sizes those buffers so the (chunk, num_classes) logits buffer matches the
input -- the efficient pick while num_classes >> in_features -- but it does
not size the (chunk, in_features) accumulator, which then grows as
chunk*in_features ~ num_batches*in_features^2/num_classes, quadratic in
in_features. So once in_features approaches num_classes the chunked peak
crosses the full-materialization reference.

This adds a 'budget' chunking method that sizes the chunk so the total
per-chunk buffer is about one input footprint, bounding it to a fixed
multiple of the input regardless of in_features; peak memory then stays
parallel to the reference instead of overtaking it. 'auto' keeps the
aspect_ratio pick below the crossing region and switches to budget above it
(in_features*acc_bytes >= num_classes*logits_bytes), so the common
num_classes >> in_features case is unchanged and only the large-in_features
regime gains the bound; an explicit aspect_ratio or budget is respected.

The default budget eps=1 ("buffer ~ one input footprint") is fixed by
continuity at the switch boundary, not by a memory objective: any
memory-only criterion is monotonic in eps and degenerates to eps=0 (one row
per chunk, slowest). eps=1 makes the budget chunk size meet aspect_ratio's
at the boundary, so the auto method change introduces no jump; 'budget:E'
overrides it.

The 2*(N+V) precompute floor is shared with the reference (and liger, which
precomputes likewise), so budget bounds the overage rather than undercutting
the reference -- the chunked path stays more accurate throughout (fp32-acc
logits vs the reference's input-dtype logits), which is why auto switches
method rather than falling back. Target-kind agnostic (index and prob).

This PR was authored with the assistance of an AI tool (Claude).

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Comment by pearu** [comment]
### A100 benchmark (bf16, index targets): `budget` bounds the large-`in_features` crossing

budget_justification_index_bfloat16_NVIDIA_A100_SXM4_80GB

Empirical confirmation of the mechanism described in the PR. Peak memory (GiB), `num_classes=32000`, `num_tokens=4096`:

| in_features | reference | auto before (aspect_ratio) | auto after (budget) |
|---|---|---|---|
| 4096 | 1.27 | 0.63 | 0.63 |
| 8192 | 1.61 | 1.26 | 1.26 |
| 16384 | 2.71 | 2.53 | 2.53 |
| 32768 | 4.91 | **5.16** | **4.80** |
| 49152 | 7.11 | **7.55** | **7.19** |

Below the crossing, chunked sits under reference and `budget == aspect_ratio` (switch not engaged). At `in_features >= 32768`, `aspect_ratio` rises above reference (5.16 > 4.91, 7.55 > 7.11) while `budget` holds the reference floor (4.80, 7.19) -- meeting reference near `in_features ~ num_classes`, i.e. bounding the overage rather than beating reference.

Larger for bf16 probability targets (extra `(chunk, num_classes)` buffer also bounded) -- see #187053.

File: torch/nn/modules/linear_cross_entropy_options.py [hunk 1]
```diff
@@ -83,9 +86,22 @@ class LinearCrossEntropyOptions:
...
+      ``in_features``; see ``_adjust``). Falls back to ``"aspect_ratio:2"``
+      for unlisted pairs.
+    - ``"budget"`` -- sizes each chunk so the per-chunk working buffers
+      occupy about one input footprint
+      (``num_batches * in_features * itemsize`` bytes). Unlike
+      ``aspect_ratio``, this bounds the buffer to a fixed multiple of the
+      input, so peak memory stays close to the reference even when
+      ``in_features`` approaches ``num_classes`` (where ``aspect_ratio``'s
+      ``(batch_chunk_size, in_features)`` accumulator grows quadratically
+      and overtakes the reference). The default for ``"auto"``.
```
Comment (slayton58) [inline torch/nn/modules/linear_cross_entropy_options.py:101]: Are there any numerics changes in changing the chunk-size? If so the default behavior has changed in this PR

File: torch/nn/modules/linear_cross_entropy_options.py [hunk 2]
```diff
@@ -148,9 +164,16 @@ def __post_init__(self):
             raise ValueError(f"invalid acc_policy: {self.acc_policy!r}")
         if self.chunking_method is not None and self.chunking_method != "auto":
```
Comment (slayton58) [inline torch/nn/modules/linear_cross_entropy_options.py:165]: Pull this out into a `_check_chunking_is_valid` or similar?

**Review by slayton58 — COMMENTED** [review]
Looks ok, with a general comment: as the number of options increases, it feels to me the code is getting a little unwieldy in the "skim over it and work out what happens sense" -- there's a bunch of:
```
if option_a:
    # do something here
    ...
    # over multiple lines
elif option_b:
    # do something else
    ...
else:
    # if all else fails, do this
```
I'm not 100% sure how to make this constructive, but I think it's worth flagging, and worth thinking about. I'm a fan of basically function tables:
```
_compute_batch_chunk_size_fns = {
    'budget': _compute_batch_chunk_size_budget,
    'default': _compute_batch_chunk_size,
}

...
compute_batch_chunk_size_fn = _compute_batch_chunk_size_fns[method]

batch_chunk_size = compute_batch_chunk_size_fn(...)
```
but I know that love is not universal :)

Not a blocking concern I feel, but I do feel compelled to raise it

**Comment by slayton58** [comment]
@claude Review this PR, exhaustively find all potential issues, concentrating on subtle problems a human reviewer might miss. Ignore small things, and concisely flag only major issues.

File: torch/nn/modules/linear_cross_entropy_options.py (hunk 1 above)
Comment (pearu) [reply torch/nn/modules/linear_cross_entropy_options.py:101]: Good question. The chunked op is chunk-size-invariant within floating-point rounding, and there's a test for exactly this: `test_linear_cross_entropy_chunk_size_invariance` runs at fp64 and asserts the forward and both gradients match across `batch_chunk_size` in {1, 2, 3, 4, 7, 13}. Summing per-chunk contributions is associative, so chunk size has no algebraic effect; at fp64 the residual is ~1 ULP/op, far under `assertEqual`'s tolerance. So the `auto` switch to `budget` above the crossing changes only *which* chunk size is resolved, not the result or its correctness -- the default's loss and gradients are unchanged up to rounding.

The one thing that varies with chunk size is mixed-precision rounding: at fp16/bf16 with fp32 accumulation, different chunk sizes select different matmul kernels and do slightly different low-precision cross-chunk accumulation, so results differ at the few-ULP level -- within the documented per-policy accuracy budget enforced by the ULP caps in `_test_linear_cross_entropy_loss`. That's exactly why the invariance test is fp64: to isolate the algorithmic invariance from matmul-kernel rounding, which would otherwise dominate at LLM-scale shapes. So there's no correctness or default-behavior change from the switch -- only rounding-level differences above the crossing region.

I'll add a sentence to the `chunking_method` docstring making this explicit.

Comment (pearu) [reply torch/nn/modules/linear_cross_entropy_options.py:101]: Why the bot's point 1 in https://github.com/pytorch/pytorch/pull/187271#issuecomment-4718845314 misses: it inferred the numerics shift from accurate's docstring line ("better input-grad accuracy with larger chunks") rather than from the invariance test - but input_grad is per-row independent (no cross-chunk accumulation), so the fp64 invariance test holds across all chunk sizes. The bot reasoned from a docstring sentence instead of the test that directly settles it.

**Comment by pearu** [comment]
Closing in favor of https://github.com/pytorch/pytorch/pull/187838
